### PR TITLE
[docs] Add distance expression entry; update other expressions

### PR DESCRIPTION
**Pull request by chloekraw**
Screen Shot 2020-05-08 at 12 34 26 AM
Screen Shot 2020-05-08 at 12 34 46 AM
Screen Shot 2020-05-08 at 12 47 47 AM
Screen Shot 2020-05-08 at 12 47 35 AM
Screen Shot 2020-05-08 at 12 35 25 AM
Screen Shot 2020-05-08 at 3 23 16 PM
Screen Shot 2020-05-08 at 3 23 20 PM

### Changelog:
- `distance`: 
   - added entry and SDK support table
- `within`: 
   - re-organized entry to increase clarity and conciseness
- `format`: 
   - mentioned the `image` expression explicitly in the documentation entry
   - re-formatted with bullet points
   - added version support information for Android SDK
- `index-of`: 
   - reworded entry to match language of `in` expression
   - removed misleading usage of "second argument," which doesn't match how current docs for [`get`](https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#get) and [`has`](https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#has) use "second argument." Considering `["index-of", needle, haystack, fromIndex]`, it's the third argument that can be optionally provided to configure the start index of the search.
- `slice`: 
   - reworded entry to match language of `in` expression
   - moved group from `String` to `Lookup` because `slice` is closer to the `in` and `index-of` expressions in function than other expressions in the `String` group. No other expressions in the `String` group work on arrays.
   - removed misleading usage of "second argument." Considering `["slice", input, startIndex, endIndex]`, it's the third argument that can be optionally provided to configure the end index.
- `match`: 
   - minor nitpick to match other entries which use `-` instead of `*` for formatting bullet points
- `interpolate`: no changes made

cc @lbutler @zmiao @alexshalamov @1ec5

### CODE REVIEW

**Comment by chloekraw** [comment]
@colleenmcginnis I tried to push my branch to our staging server but it didn't work. Not sure if it's possible for this part of our documentation?

File: src/style-spec/reference/v8.json [hunk 1]
```diff
@@ -3345,6 +3355,17 @@
           }
         }
       },
+      "distance": {
+        "doc": "Returns the shortest distance in meters between the evaluated feature and the input geometry. The evaluated feature and input value can be a valid GeoJSON of type `Point`, `MultiPoint`, `LineString`, `MultiLineString`, `Polygon`, `MultiPolygon`, `Feature`, or `FeatureCollection`. Results at zoom level 12 and below may vary in precision due to precision loss in conversions between tile coordinates and `[Longitude, Latitude]` at lower zoom levels.",
```
Comment (chloekraw) [inline src/style-spec/reference/v8.json:3359]: > Results at zoom level 12 and below may vary in precision due to precision loss in conversions between tile coordinates and `[Longitude, Latitude]` at lower zoom levels."

Ah, I could simplify this part to "[encoding geometries](https://docs.mapbox.com/vector-tiles/specification/#encoding-geometry)". Maybe something like:

Distance values returned at zoom level 12 and below may vary in precision due to loss in precision from encoding geometries.

**Comment by colleenmcginnis** [comment]
>@colleenmcginnis I tried to push my branch to our staging server but it didn't work. Not sure if it's possible for this part of our documentation?

It's possible, but it's not something we do often. It would require updating the submodule in /mapbox-gl-js-docs to this commit and then pushing to `publisher-staging` from there. As far as I know, this repo is no longer connected to publisher at all.

If you were just pushing to staging to check the formatting, that's something I (or likely someone from the GL JS team) could do as a part of a PR review. In the future we could pair on something like this, but unfortunately this week is a busy one for our team so if this is an urgent change I'd recommend asking for a reviewer to check formatting locally instead.

File: src/style-spec/reference/v8.json [hunk 2]
```diff
@@ -2816,7 +2825,7 @@
         }
       },
       "format": {
-        "doc": "Returns `formatted` text containing annotations for use in mixed-format `text-field` entries. For a `text-field` entries of a string type, following option object's properties are supported: If set, the `text-font` value overrides the font specified by the root layout properties. If set, the `font-scale` value specifies a scaling factor relative to the `text-size` specified in the root layout properties. If set, the `text-color` value overrides the color specified by the root paint properties for this layer.",
+        "doc": "Returns a `formatted` string for use in the `text-field` property. The input may contain a plain `string`, an `image` provided through the [`'image'`](#types-image) expression, or a combination  of the two. Supported styling options:\n- `\"text-font\"`: Overrides the font stack specified by the root layout property.\n- `\"text-color\"`: Overrides the color specified by the root paint property.\n- `\"font-scale\"`: Applies the specified a scaling factor relative to the `text-size` specified by the root layout property.",
```
Comment (asheemmamoowala) [inline src/style-spec/reference/v8.json:2828]: Removing the `mixed-format` makes it harder to understand the purpose of this expression type and when to use it. It's also not clear from the description how to pair the sections with their override options.

```suggestion
        "doc": "Returns a `formatted` string for displaying mixed-format text in the `text-field` property. The input may contain a string literal or expression, or an  [`'image'`](#types-image) expression, or a combination  of the two. Strings can be followed with a style override object that supports the following properties:\n- `\"text-font\"`: Overrides the font stack specified by the root layout property.\n- `\"text-color\"`: Overrides the color specified by the root paint property.\n- `\"font-scale\"`: Applies the specified number as a scaling factor relative to the `text-size` specified by the root layout property.",
```

File: src/style-spec/reference/v8.json (hunk 1 above)
Comment (asheemmamoowala) [inline src/style-spec/reference/v8.json:3359]: > The evaluated feature and input value can be a valid GeoJSON of type

Does this imply that the evaluated feature cannot be part of a vector tile?

> `Point`, `MultiPoint`, `LineString`, `MultiLineString`, `Polygon`, `MultiPolygon`, `Feature`, or `FeatureCollection`

Include details on what distance is measured among some of the pairs and what happens withe `Multi*` and `FeatureCollection` types?

cc @zmiao

File: src/style-spec/reference/v8.json [hunk 3]
```diff
@@ -3345,6 +3355,17 @@
           }
         }
       },
+      "distance": {
+        "doc": "Returns the shortest distance in meters between the evaluated feature and the input geometry. The evaluated feature and input value can be a valid GeoJSON of type `Point`, `MultiPoint`, `LineString`, `MultiLineString`, `Polygon`, `MultiPolygon`, `Feature`, or `FeatureCollection`. Results at zoom level 12 and below may vary in precision due to precision loss in conversions between tile coordinates and `[Longitude, Latitude]` at lower zoom levels.",
+        "group": "Math",
```
Comment (asheemmamoowala) [inline src/style-spec/reference/v8.json:3360]: What do you think of adding a new `Geospatial` group for `within` and `distance`?

File: src/style-spec/reference/v8.json (hunk 1 above)
Comment (chloekraw) [reply src/style-spec/reference/v8.json:3359]: Hmm, good point, it shouldn't imply that. I'm open to better suggestions! 

> Include details on what distance is measured among some of the pairs and what happens withe Multi* and FeatureCollection types?

Yeah, good idea.

File: src/style-spec/reference/v8.json (hunk 3 above)
Comment (chloekraw) [reply src/style-spec/reference/v8.json:3360]: I think we can always do this later if we find ourselves adding more geospatial expressions. Generally I'm hesitant to create a new category for fewer than 3 members, and `within` fits in the Decision group pretty well.  I agree, `distance` is a bit more awkward in the Math group, but I think it works fine enough.

File: src/style-spec/reference/v8.json (hunk 1 above)
Comment (chloekraw) [reply src/style-spec/reference/v8.json:3359]: @zmiao for the input data, is what I wrote correct about what’s supported?

Yeah, maybe omitting “evaluated” feature from the description could work. What do you think about Asheem’s suggestion to add more detail on which distance is measured and the behavior for Multi* types? Maybe you could provide the info here and we could evaluate whether it’s worth including

**Comment by asheemmamoowala** [comment]
Noting here that the the unit test failure is from a missing implementation:
```
test/unit/style-spec/expression.test.js .............. 9/10
  v8.json includes all definitions from style-spec
  not ok should be equivalent
    --- wanted
    +++ found
       "coalesce"
       "collator"
       "concat"
       "cos"
    -  "distance"
       "downcase"
       "e"
       "feature-state"
       "floor"
    at:
      line: 14
      column: 7
      file: test/unit/style-spec/expression.test.js
      type: Test
```

File: src/style-spec/reference/v8.json (hunk 1 above)
Comment (chloekraw) [reply src/style-spec/reference/v8.json:3359]: @zmiao @asheemmamoowala which do you think is better?

Comment (asheemmamoowala) [reply src/style-spec/reference/v8.json:3359]: > Distance values returned at zoom level 12 and below may vary in precision due to loss in precision from encoding geometries.

☝️

Comment (zmiao) [reply src/style-spec/reference/v8.json:3359]: > for the input data, is what I `wrote` correct about what’s supported?
Yes, it is correct.

> What do you think about Asheem’s suggestion to add more detail on which distance is measured for different combinations and the behavior for Multi* types? 

I think it is a good idea to explain it, I will provide the details soon.

**Comment by chloekraw** [comment]
@mapbox/gl-js I added screenshots previewing all of the expressions doc updates in the first post of this ticket. This PR pairs with [this branch](https://github.com/mapbox/mapbox-gl-js-docs/compare/ck-update-expressions-metadata) in mapbox-gl-js-docs. 

@ryanhamley I wasn't sure about the "needle / haystack" terminology in the [`in` expression metadata](https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#in). It seems much more specific than other terms that we use -- I think when values can be multiple types we previously used generic terms like `InputType` and `OutputType` to stay away from non-generalizable details that are specific enough that they can (and do) serve as the variable names in the implementation.

It's a tough trade-off between clarity and consistency. I tried the version below first and I do think it's too messy and unclear with the multiple `|` and `,` so I'm thinking about changing it to `["in", InputType (boolean, string, or number), InputType (array or string)]`. Thoughts?

Screen Shot 2020-05-08 at 12 39 16 AM

Screen Shot 2020-05-08 at 12 35 30 AM

**Comment by ryanhamley** [comment]
@chloekraw I think if that's what we've standardized the rest of the docs around then it's ok to make it more generic. I definitely prefer the `["in", InputType (boolean, string, or number), InputType (array or string)]` formulation to be clearer about the grouping of types. I think the only concern is that it may become confusing about which argument is which. That's partially what the `needle` and `haystack` nomenclature represents, but we could think of other ways to be more explicit about this distinction. Or maybe the implicit recognition by type is sufficient. It's explicitly documented as finding a value in a string or array and only the second argument supports those types.

TLDR: I'm not married to the `needle/haystack` terminology. It was mostly just to clarify the order of arguments.

**Comment by chloekraw** [comment]
@ryanhamley thanks! yeah, I think that makes sense, and I do appreciate that your version is clearer about the usage. I think we can also achieve that goal through creating a GL JS example. I was trying to think of a way to add a sentence to the expression definition that could clarify the first vs. second argument, but I came up blank. I think the first argument could be described as a e.g. "search query" but I really couldn't think of anything for the second argument. If you have a suggestion, let me know!

**Comment by chloekraw** [comment]
@asheemmamoowala @ryanhamley @ansis @1ec5 @zmiao @alexshalamov thanks for all the feedback! I'm going to merge this and cherrypick to `release-danube` now, but happy to revisit any of the changes in the future.
